# Phase 4 Project - Time Series Modeling 
By Kristen Davis 

## 1 - Notebook Set Up 

In [110]:
#import nessesary libraries 
import pandas as pd
import pandas.tseries  

import numpy as np  

import statsmodels.api as sm 

import matplotlib.pyplot as plt  
import plotly.figure_factory as ff 
import plotly.graph_objects as go
import seaborn as sns 

In [111]:
#load the data from teh zillow data set
df = pd.read_csv('zillow_data.csv') 

## 2 - EDA 

In [112]:
#preview the dataframe
df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
0         1  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800   
1         2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
2         3  210400.0  212200.0  212200.0  ...   321000   320600   320200   
3         4  498100.0  500900.0  503100.0  ...  1289800  1287700  1287400   
4         5   77300.0   77300.0   77300.0  ...   119100   119400   120000   

   2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1009600  1013300  1018700  1024400  1030700  1033800  1030600  
1   314100   315000   316600   318100   319600   321100   321800  
2   320400   320800   321200   321200   323000   326900   329900  
3  1291500  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120300   120500   121000   121500  

[5 rows x 272 columns]

In [113]:
#dataframe metadata
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14723 entries, 0 to 14722
Columns: 272 entries, RegionID to 2018-04
dtypes: float64(219), int64(49), object(4)
memory usage: 30.6+ MB


In [114]:
df.columns.to_list()

['RegionID',
 'RegionName',
 'City',
 'State',
 'Metro',
 'CountyName',
 'SizeRank',
 '1996-04',
 '1996-05',
 '1996-06',
 '1996-07',
 '1996-08',
 '1996-09',
 '1996-10',
 '1996-11',
 '1996-12',
 '1997-01',
 '1997-02',
 '1997-03',
 '1997-04',
 '1997-05',
 '1997-06',
 '1997-07',
 '1997-08',
 '1997-09',
 '1997-10',
 '1997-11',
 '1997-12',
 '1998-01',
 '1998-02',
 '1998-03',
 '1998-04',
 '1998-05',
 '1998-06',
 '1998-07',
 '1998-08',
 '1998-09',
 '1998-10',
 '1998-11',
 '1998-12',
 '1999-01',
 '1999-02',
 '1999-03',
 '1999-04',
 '1999-05',
 '1999-06',
 '1999-07',
 '1999-08',
 '1999-09',
 '1999-10',
 '1999-11',
 '1999-12',
 '2000-01',
 '2000-02',
 '2000-03',
 '2000-04',
 '2000-05',
 '2000-06',
 '2000-07',
 '2000-08',
 '2000-09',
 '2000-10',
 '2000-11',
 '2000-12',
 '2001-01',
 '2001-02',
 '2001-03',
 '2001-04',
 '2001-05',
 '2001-06',
 '2001-07',
 '2001-08',
 '2001-09',
 '2001-10',
 '2001-11',
 '2001-12',
 '2002-01',
 '2002-02',
 '2002-03',
 '2002-04',
 '2002-05',
 '2002-06',
 '2002-07',
 '2

In [115]:
df.isna().sum().to_list()

[0,
 0,
 0,
 0,
 1043,
 0,
 0,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1039,
 1038,
 1038,
 1038,
 1038,
 1038,
 1038,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 1036,
 918,
 918,
 918,
 918,
 918,
 918,
 887,
 887,
 887,
 887,
 887,
 887,
 866,
 866,
 866,
 866,
 866,
 866,
 814,
 814,
 801,
 801,
 801,
 801,
 723,
 723,
 723,
 723,
 723,
 723,
 667,
 667,
 667,
 667,
 667,
 667,
 640,
 640,
 640,
 640,
 640,
 640,
 620,
 620,
 620,
 620,
 620,
 620,
 613,
 613,
 613,
 613,
 613,
 613,
 607,
 607,
 607,
 607,
 607,
 607,


In [116]:
#States represented by the data 
df['State'].unique()

array(['IL', 'TX', 'NY', 'CA', 'FL', 'TN', 'NC', 'GA', 'DC', 'MO', 'OK',
       'AZ', 'NJ', 'MD', 'VA', 'WA', 'OH', 'MI', 'MA', 'KS', 'NM', 'CT',
       'NV', 'PA', 'CO', 'OR', 'IN', 'SC', 'KY', 'AR', 'ND', 'MN', 'AL',
       'DE', 'LA', 'MS', 'ID', 'MT', 'HI', 'WI', 'UT', 'ME', 'SD', 'WV',
       'IA', 'RI', 'NE', 'WY', 'AK', 'NH', 'VT'], dtype=object)

In [117]:
#number of region ids by state
df.groupby('State').count()

RegionID  RegionName  City  Metro  CountyName  SizeRank  1996-04  \
State                                                                     
AK           28          28    28     22          28        28       28   
AL          183         183   183    175         183       183      158   
AR          105         105   105    100         105       105       97   
AZ          230         230   230    228         230       230      215   
CA         1224        1224  1224   1182        1224      1224     1188   
CO          249         249   249    229         249       249      247   
CT          124         124   124    124         124       124      124   
DC           18          18    18     18          18        18       18   
DE           41          41    41     41          41        41       26   
FL          785         785   785    758         785       785      765   
GA          345         345   345    317         345       345      291   
HI           62          62    62     62          62        62       51   
IA          158         158   158    156         158       158      148   
ID          110         110   110     87         110       110       95   
IL          547         547   547    537         547       547      523   
IN          428         428   428    363         428       428      392   
KS          241         241   241    184         241       241      168   
KY          139         139   139    132         139       139      120   
LA          193         193   193    173         193       193      145   
MA          417         417   417    416         417       417      414   
MD          317         317   317    300         317       317      313   
ME           86          86    86     85          86        86       77   
MI          499         499   499    457         499       499      453   
MN          375         375   375    339         375       375      344   
MO          319         319   319    269         319       319      269   
MS          153         153   153     98         153       153       67   
MT           71          71    71     43          71        71       52   
NC          428         428   428    391         428       428      402   
ND           31          31    31     21          31        31        0   
NE           83          83    83     78          83        83       70   
NH          199         199   199    178         199       199      195   
NJ          502         502   502    502         502       502      502   
NM           60          60    60     55          60        60       53   
NV          103         103   103    103         103       103       88   
NY         1015        1015  1015    963        1015      1015      985   
OH          588         588   588    570         588       588      529   
OK          221         221   221    200         221       221      210   
OR          224         224   224    216         224       224      210   
PA          831         831   831    816         831       831      797   
RI           59          59    59     59          59        59       59   
SC          206         206   206    199         206       206      192   
SD           19          19    19     19          19        19       10   
TN          404         404   404    336         404       404      399   
TX          989         989   989    873         989       989      944   
UT          121         121   121    109         121       121      116   
VA          401         401   401    365         401       401      372   
VT           16          16    16     16          16        16       16   
WA          341         341   341    323         341       341      336   
WI          332         332   332    302         332       332      327   
WV           72          72    72     71          72        72       60   
WY           31          31    31     20          31        31       24   

       1996-05  1996-06  1996-0

This data set contains:  
- 14722 enteries representing data on single family residential home sales, between 1996 & 2018  
- The following geographical identifier columns 'RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', & 'SizeRank' 
- Average home sale prices each month between the years 1996 & 2018
- Missing values in the 'Metro' identifier & a varying number of missing values between the years 1996 - 2014 (all rows missing values are less than 1,500 representing less than 10% of the data)
- Entries from all 50 states  
- A varying number of region ids represented for each state

## 3 - Data Preprocessing 

InvestTex the client I am building a recomendation system for is interested in investing only in the Dallas/ Fort Worth Metroplex and thus I will need to limit to scope of my data. 

In [118]:
#drop rows with NAN values 
df.dropna(inplace=True)

In [119]:
#listings in Texas
texas = df.loc[df['State'] == 'TX']

In [120]:
#listings in Dallas
dallas = texas.loc[texas['Metro'] == "Dallas-Fort Worth"] 
#preview data 
dallas

RegionID  RegionName           City State              Metro  \
1         90668       75070       McKinney    TX  Dallas-Fort Worth   
34        90654       75052  Grand Prairie    TX  Dallas-Fort Worth   
52        90637       75034         Frisco    TX  Dallas-Fort Worth   
54        90823       75287         Dallas    TX  Dallas-Fort Worth   
69        90795       75243         Dallas    TX  Dallas-Fort Worth   
...         ...         ...            ...   ...                ...   
12567     90892       75422       Campbell    TX  Dallas-Fort Worth   
12598     90893       75423        Celeste    TX  Dallas-Fort Worth   
13271     91239       76093      Rio Vista    TX  Dallas-Fort Worth   
13312     91179       76005      Arlington    TX  Dallas-Fort Worth   
13922     91222       76064       Maypearl    TX  Dallas-Fort Worth   

      CountyName  SizeRank   1996-04   1996-05   1996-06  ...  2017-07  \
1         Collin         2  235700.0  236900.0  236700.0  ...   308000   
34        Dallas        35  100800.0  100900.0  101000.0  ...   186700   
52        Collin        53  285200.0  287600.0  289500.0  ...   408700   
54        Dallas        55  184400.0  183300.0  182400.0  ...   319600   
69        Dallas        70  136500.0  136400.0  136400.0  ...   279900   
...          ...       ...       ...       ...       ...  ...      ...   
12567       Hunt     12568   74000.0   74200.0   74500.0  ...   105400   
12598       Hunt     12599   55700.0   56100.0   56400.0  ...    84000   
13271    Johnson     13272   70000.0   70200.0   70300.0  ...   137100   
13312    Tarrant     13313  200100.0  200000.0  199800.0  ...   388700   
13922      Ellis     13923   83800.0   84000.0   84100.0  ...   163100   

       2017-08  2017-09  2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  \
1       310000   312500   314100   315000   316600   318100   319600   321100   
34      188200   189700   191400   193000   194300   195200   196700   199100   
52      411000   414300   418000   421600   423000   422800   424000   427600   
54      322300   325000   325900   326200   326700   328000   330800   333900   
69      280900   280900   281700   283300   286400   290200   293600   295600   
...        ...      ...      ...      ...      ...      ...      ...      ...   
12567   107500   108900   109500   110300   111600   112700   113900   115400   
12598    84400    85500    86400    87700    88300    88700    88800    89000   
13271   135400   135900   137700   137400   137200   138200   141300   145100   
13312   386400   383300   379400   375100   374200   374200   373700   373900   
13922   165100   170600   173500   172300   170500   169400   169300   171300   

       2018-04  
1       321800  
34      201100  
52      430200  
54      335100  
69      296100  
...        ...  
12567   117200  
12598    89000  
13271   147600  
13312   373900  
13922   174200  

[207 rows x 272 columns]

In [121]:
#identify unique counties to find FIP codes 
dallas['CountyName'].unique()

array(['Collin', 'Dallas', 'Denton', 'Johnson', 'Tarrant', 'Kaufman',
       'Ellis', 'Rockwall', 'Hood', 'Parker', 'Hunt', 'Somervell'],
      dtype=object)

In [122]:
#visualize the dallas /ft worth metroplex
fips = ['48085', '48113', '48121', '48251', '48439', '48257', '48139', '48397', '48221', '48367', '48231', '48425']
values = range(len(fips)) 

fig = ff.create_choropleth(fips=fips, values=values, scope=['Texas'], 
                           county_outline={'color': 'rgb(15, 15, 55)', 'width': 0.5}, 
                           state_outline={'color': 'rgb(15, 15, 55)', 'width': 1},
                           title='Counties of Investment Interest')
fig.layout.template = None
fig.show() 

In [123]:
#get the number of zipcodes represented by County Name 
dallas.groupby('CountyName').count()

RegionID  RegionName  City  State  Metro  SizeRank  1996-04  \
CountyName                                                                
Collin            23          23    23     23     23        23       23   
Dallas            69          69    69     69     69        69       69   
Denton            24          24    24     24     24        24       24   
Ellis              6           6     6      6      6         6        6   
Hood               2           2     2      2      2         2        2   
Hunt               8           8     8      8      8         8        8   
Johnson            8           8     8      8      8         8        8   
Kaufman            2           2     2      2      2         2        2   
Parker             2           2     2      2      2         2        2   
Rockwall           3           3     3      3      3         3        3   
Somervell          1           1     1      1      1         1        1   
Tarrant           59          59    59     59     59        59       59   

            1996-05  1996-06  1996-07  ...  2017-07  2017-08  2017-09  \
CountyName                             ...                              
Collin           23       23       23  ...       23       23       23   
Dallas           69       69       69  ...       69       69       69   
Denton           24       24       24  ...       24       24       24   
Ellis             6        6        6  ...        6        6        6   
Hood              2        2        2  ...        2        2        2   
Hunt              8        8        8  ...        8        8        8   
Johnson           8        8        8  ...        8        8        8   
Kaufman           2        2        2  ...        2        2        2   
Parker            2        2        2  ...        2        2        2   
Rockwall          3        3        3  ...        3        3        3   
Somervell         1        1        1  ...        1        1        1   
Tarrant          59       59       59  ...       59       59       59   

            2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
CountyName                                                                 
Collin           23       23       23       23       23       23       23  
Dallas           69       69       69       69       69       69       69  
Denton           24       24       24       24       24       24       24  
Ellis             6        6        6        6        6        6        6  
Hood              2        2        2        2        2        2        2  
Hunt              8        8        8        8        8        8        8  
Johnson           8        8        8        8        8        8        8  
Kaufman           2        2        2        2        2        2        2  
Parker            2        2        2        2        2        2        2  
Rockwall          3        3        3        3        3        3        3  
Somervell         1        1        1        1        1        1        1  
Tarrant          59       59       59       59       59       59       59  

[12 rows x 271 columns]

In [124]:
labels = ['Collin', 'Dallas', 'Denton', 'Johnson', 'Tarrant', 'Kaufman',
       'Ellis', 'Rockwall', 'Hood', 'Parker', 'Hunt', 'Somervell']
values = [23, 69, 24, 8, 59, 2, 6, 3, 2, 2, 8, 1]

fig = go.Figure(data=[go.Pie(labels=labels, values=values, textinfo='percent',
                             insidetextorientation='radial'
                            )])
fig.show()

In [125]:
#list out zipcodes in the metroplex
dallas['RegionName'].unique()

array([75070, 75052, 75034, 75287, 75243, 75067, 76028, 76063, 75228,
       76244, 75056, 75217, 75206, 75035, 75150, 76116, 75043, 75002,
       75211, 75204, 76179, 76137, 76051, 75098, 75115, 75093, 75040,
       75007, 75149, 75061, 75025, 75231, 76133, 75068, 75080, 75071,
       76010, 75074, 75219, 75023, 75024, 75227, 75063, 75006, 75062,
       76112, 76017, 75214, 75248, 76039, 75104, 76108, 75019, 75126,
       75165, 75050, 75013, 76021, 75044, 75028, 76107, 76119, 75060,
       76180, 75075, 76262, 75154, 75081, 75220, 76210, 75069, 75038,
       76248, 75087, 76006, 76013, 75077, 76065, 75230, 76132, 76131,
       75252, 76227, 75238, 75051, 75033, 76049, 76053, 75234, 75254,
       75042, 75229, 75010, 76182, 76117, 76040, 76016, 75039, 76048,
       76087, 75032, 76014, 76020, 76012, 75224, 75218, 76033, 75241,
       75089, 76001, 76109, 76011, 76123, 76201, 75189, 76209, 75232,
       75205, 76226, 75041, 76140, 76092, 75225, 76036, 75160, 76002,
       76114, 76126,

There are still 207 rows of housing data that fall within the Dallas/ Fort Worth - one of the defined metrics I will be using to recommend the best investment zipcodes are ones that fall within good school districts. School district preformance is a powerful predictor of housing value so I will group zipcodes by school district and then select only those zipcodes that fall into high preforming school districts (A-/A/A+)

In [126]:
#visualize distribution of school district grades in the metroplex  
grades=['A+', 'A', 'A-', 'B+', 'B', 'B-', 'C+', 'C', 'C-'] 

fig = go.Figure([go.Bar(x=grades, y=[17, 36, 22, 30, 26, 48, 17, 8, 1])]) 
fig.update_layout(
    title='Frequency Listing by of School District Grade',
    xaxis=dict(
        title='School District Grade',
        titlefont_size=16,
        tickfont_size=14,
    ),
    yaxis=dict(
        title='Number of Zipcodes',
        titlefont_size=16,
        tickfont_size=14,
    ),
    legend=dict(
        x=0,
        y=1.0,
        bgcolor='rgba(255, 255, 255, 0)',
        bordercolor='rgba(255, 255, 255, 0)'
    ),
    barmode='group',
    bargap=0.15, # gap between bars of adjacent location coordinates.
    bargroupgap=0.1 # gap between bars of the same location coordinate.
)
fig.show()

In [127]:
#slice zipcodes in school districts with A+/A/A-  
low_zipcodes = [76227, 76020, 76135, 76028, 75287, 75007, 75006, 75234, 75423, 75119, 75126, 75043, 75040, 75044,
                75042, 75089, 75041, 75088, 76059, 75065, 75453, 75407, 76259, 75243, 75080, 75248, 75081, 75238, 
                75496, 76010, 76012, 76005, 76006, 76013, 76016, 76011, 76014, 76015, 76018, 75442, 76049, 76048, 
                75050, 75051, 76017, 76060, 75068, 75150, 75149, 75180, 75181, 75253, 75189, 76087, 76108, 75409, 
                75228, 75217, 75206, 75211, 75204, 75231, 75219, 75227, 75214, 75141, 75224, 75218, 75241, 75001, 
                75201, 75220, 75230, 75254, 75229, 75235, 75237, 75209, 75159, 75244, 75233, 75232, 76116, 76133, 
                76112, 76107, 76119, 76126, 76120, 76105, 76111, 76104, 76103, 76164, 76109, 75061, 75062, 75060, 
                75038, 75039, 76249, 76258, 76266, 76009, 76033, 75428, 75173, 75166, 75137, 75249, 75116, 75236, 
                76140, 76134, 75402, 75146, 75134, 76064, 75474, 75160, 75422, 76114, 75104, 76132, 76123, 76036, 
                76093, 76084, 75115]  

for index, row in dallas.iterrows():
    if row['RegionName'] in low_zipcodes:
        dallas.drop(index, inplace=True) 
        
dallas.head()

RegionID  RegionName           City State              Metro CountyName  \
1       90668       75070       McKinney    TX  Dallas-Fort Worth     Collin   
34      90654       75052  Grand Prairie    TX  Dallas-Fort Worth     Dallas   
52      90637       75034         Frisco    TX  Dallas-Fort Worth     Collin   
81      90665       75067     Lewisville    TX  Dallas-Fort Worth     Denton   
153     91221       76063      Mansfield    TX  Dallas-Fort Worth    Tarrant   

     SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
1           2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
34         35  100800.0  100900.0  101000.0  ...   186700   188200   189700   
52         53  285200.0  287600.0  289500.0  ...   408700   411000   414300   
81         82  118400.0  118500.0  118700.0  ...   220400   223800   227600   
153       154  148100.0  148000.0  147800.0  ...   257200   258700   259000   

     2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
1     314100   315000   316600   318100   319600   321100   321800  
34    191400   193000   194300   195200   196700   199100   201100  
52    418000   421600   423000   422800   424000   427600   430200  
81    230600   233100   234900   235900   237700   240300   241800  
153   259600   260200   261100   262200   264100   266500   267800  

[5 rows x 272 columns]

After applying the first metric for what the best zipcodes in the area would be to invest in we are left with 78 zipcodes that fall into high preforming school districts. 

In [128]:
#Dataframe of zip/ school district & grade
df2 = pd.DataFrame(np.array([['A+', 75002, 'Allen ISD'], ['A+', 75013, 'Allen ISD'], ['A+', 76226, 'Argyle ISD'],
                            ['A+', 76092, 'Carroll ISD'], ['A+', 75063, 'Coppel ISD'], ['A+', 75019, 'Coppel ISD'],
                            ['A+', 75035, 'Frisco ISD'], ['A+', 75024, 'Frisco ISD'], ['A+', 75033, 'Frisco ISD'], 
                            ['A+', 76051, 'Grapevine-Colleyville ISD'], ['A+', 76034, 'Grapevine-Colleyville ISD'],
                            ['A+', 75205, 'Highland Park ISD'], ['A+', 75225, 'Highland Park ISD'], 
                            ['A+', 75185, 'Sunnyvale ISD'], ['A+', 75098, 'Wylie ISD'], ['A+', 75048, 'Wylie ISD'], 
                            ['A+', 75094, 'Wylie ISD'], ['A', 76008, 'Aledo ISD'], ['A', 76043, 'Glen Rose ISD'],
                            ['A', 76050, 'Grandview ISD'], ['A', 76039, 'Hurst-Euless-Bedford ISD'], 
                            ['A', 76021, 'Hurst-Euless-Bedford ISD'], ['A', 76053, 'Hurst-Euless-Bedford ISD'], 
                            ['A', 76040, 'Hurst-Euless-Bedford ISD'], ['A', 76054, 'Hurst-Euless-Bedford ISD'], 
                            ['A', 76029, 'Hurst-Euless-Bedford ISD'], ['A', 76244, 'Keller ISD'],
                            ['A', 76248, 'Keller ISD'], ['A', 75067, 'Lewisville ISD'], 
                            ['A', 75066, 'Lewisville ISD'], ['A', 75028, 'Lewisville ISD'], 
                            ['A', 75077, 'Lewisville ISD'], ['A', 75010, 'Lewisville ISD'], 
                            ['A', 75022, 'Lewisville ISD'],['A', 75057, 'Lewisville ISD'], 
                            ['A', 76063, 'Mansfield ISD'], ['A', 76002, 'Mansfield ISD'], 
                            ['A', 76001, 'Mansfield ISD'], ['A', 75071, 'McKinney ISD'], 
                            ['A', 75069, 'McKinney ISD'], ['A', 76262, 'Northwest ISD'], 
                            ['A', 76052, 'Northwest ISD'], ['A', 75074, 'Northwest ISD'],
                            ['A', 75023, 'Northwest ISD'], ['A', 75075, 'Northwest ISD'], 
                            ['A', 75252, 'Northwest ISD'], ['A', 75082, 'Northwest ISD'], 
                            ['A', 75087, 'Rockwall ISD'], ['A', 75032, 'Rockwall ISD'], 
                            ['A-', 76137, 'Birdville ISD'], ['A-', 76180, 'Birdville ISD'], 
                            ['A-', 76182, 'Birdville ISD'], ['A-', 76117, 'Birdville ISD'], 
                            ['A-', 76118, 'Birdville ISD'], ['A-', 76148, 'Birdville ISD'], 
                            ['A-', 75135, 'Caddo Mills ISD'], ['A-', 75009, 'Celina ISD'], 
                            ['A-', 76210, 'Denton ISD'], ['A-', 76201, 'Denton ISD'],
                            ['A-', 76208, 'Denton ISD'], ['A-', 76209, 'Denton ISD'], ['A-', 76205, 'Denton ISD'],
                            ['A-', 76207, 'Denton ISD'], ['A-', 76179, 'Eagle Mountain-Saginaw ISD'],
                            ['A-', 76131, 'Eagle Mountain-Saginaw ISD'], ['A-', 76058, 'Joshua ISD'], 
                            ['A-', 75454, 'Melissa ISD'], ['A-', 76065, 'Midlothian ISD'], 
                            ['A-', 75154, 'Red Oak ISD'], ['A-', 75165, 'Waxahachie ISD'], 
                            ['A-', 75167, 'Waxahachie ISD']]),
                   columns=['Grade', 'Zipcode', 'SchoolDistrict'])
df2

Grade Zipcode  SchoolDistrict
0     A+   75002       Allen ISD
1     A+   75013       Allen ISD
2     A+   76226      Argyle ISD
3     A+   76092     Carroll ISD
4     A+   75063      Coppel ISD
..   ...     ...             ...
66    A-   75454     Melissa ISD
67    A-   76065  Midlothian ISD
68    A-   75154     Red Oak ISD
69    A-   75165  Waxahachie ISD
70    A-   75167  Waxahachie ISD

[71 rows x 3 columns]

To further target inital analysis I will apply a second metric for success which is to look at school districts with multiple zipcodes within them. This will allow for comparison of home value accross several zipcodes in the same district to potentially identify zipcodes that have been primed for investment. 

In [129]:
#create a dataframe that only inlcudes the multi zip high preformers on school dataframe 
multi_zipcodes = ['76226', '76092', '75182', '76008', '76043', '76050', '75135', '75009', '76058', '75454', 
                  '76065', '75154']  

for index, row in df2.iterrows():
    if row['Zipcode'] in multi_zipcodes:
        df2.drop(index, inplace=True) 
        
df2

Grade Zipcode              SchoolDistrict
0     A+   75002                   Allen ISD
1     A+   75013                   Allen ISD
4     A+   75063                  Coppel ISD
5     A+   75019                  Coppel ISD
6     A+   75035                  Frisco ISD
7     A+   75024                  Frisco ISD
8     A+   75033                  Frisco ISD
9     A+   76051   Grapevine-Colleyville ISD
10    A+   76034   Grapevine-Colleyville ISD
11    A+   75205           Highland Park ISD
12    A+   75225           Highland Park ISD
13    A+   75185               Sunnyvale ISD
14    A+   75098                   Wylie ISD
15    A+   75048                   Wylie ISD
16    A+   75094                   Wylie ISD
20     A   76039    Hurst-Euless-Bedford ISD
21     A   76021    Hurst-Euless-Bedford ISD
22     A   76053    Hurst-Euless-Bedford ISD
23     A   76040    Hurst-Euless-Bedford ISD
24     A   76054    Hurst-Euless-Bedford ISD
25     A   76029    Hurst-Euless-Bedford ISD
26     A   76244                  Keller ISD
27     A   76248                  Keller ISD
28     A   75067              Lewisville ISD
29     A   75066              Lewisville ISD
30     A   75028              Lewisville ISD
31     A   75077              Lewisville ISD
32     A   75010              Lewisville ISD
33     A   75022              Lewisville ISD
34     A   75057              Lewisville ISD
35     A   76063               Mansfield ISD
36     A   76002               Mansfield ISD
37     A   76001               Mansfield ISD
38     A   75071                McKinney ISD
39     A   75069                McKinney ISD
40     A   76262               Northwest ISD
41     A   76052               Northwest ISD
42     A   75074               Northwest ISD
43     A   75023               Northwest ISD
44     A   75075               Northwest ISD
45     A   75252               Northwest ISD
46     A   75082               Northwest ISD
47     A   75087                Rockwall ISD
48     A   75032                Rockwall ISD
49    A-   76137               Birdville ISD
50    A-   76180               Birdville ISD
51    A-   76182               Birdville ISD
52    A-   76117               Birdville ISD
53    A-   76118               Birdville ISD
54    A-   76148               Birdville ISD
57    A-   76210                  Denton ISD
58    A-   76201                  Denton ISD
59    A-   76208                  Denton ISD
60    A-   76209                  Denton ISD
61    A-   76205                  Denton ISD
62    A-   76207                  Denton ISD
63    A-   76179  Eagle Mountain-Saginaw ISD
64    A-   76131  Eagle Mountain-Saginaw ISD
69    A-   75165              Waxahachie ISD
70    A-   75167              Waxahachie ISD

In [130]:
#create a dataframe that only inlcudes the multi zip high preformers on main dataframe 
multi_zipcodes = [76226, 76092, 75182, 76008, 76043, 76050, 75135, 75009, 76058, 75454, 76065, 75154]  

for index, row in dallas.iterrows():
    if row['RegionName'] in multi_zipcodes:
        dallas.drop(index, inplace=True) 
        
dallas

RegionID  RegionName           City State              Metro CountyName  \
1        90668       75070       McKinney    TX  Dallas-Fort Worth     Collin   
34       90654       75052  Grand Prairie    TX  Dallas-Fort Worth     Dallas   
52       90637       75034         Frisco    TX  Dallas-Fort Worth     Collin   
81       90665       75067     Lewisville    TX  Dallas-Fort Worth     Denton   
153      91221       76063      Mansfield    TX  Dallas-Fort Worth    Tarrant   
...        ...         ...            ...   ...                ...        ...   
7008     91310       76207         Denton    TX  Dallas-Fort Worth     Denton   
7144     90658       75057     Lewisville    TX  Dallas-Fort Worth     Denton   
7426     91328       76247     Fort Worth    TX  Dallas-Fort Worth    Tarrant   
7748     91215       76054          Hurst    TX  Dallas-Fort Worth    Tarrant   
9171     90743       75167     Waxahachie    TX  Dallas-Fort Worth      Ellis   

      SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
1            2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
34          35  100800.0  100900.0  101000.0  ...   186700   188200   189700   
52          53  285200.0  287600.0  289500.0  ...   408700   411000   414300   
81          82  118400.0  118500.0  118700.0  ...   220400   223800   227600   
153        154  148100.0  148000.0  147800.0  ...   257200   258700   259000   
...        ...       ...       ...       ...  ...      ...      ...      ...   
7008      7009  123900.0  124300.0  124600.0  ...   229600   231400   232600   
7144      7145   84000.0   84300.0   84700.0  ...   172500   175400   177100   
7426      7427  128600.0  128800.0  129100.0  ...   199500   201800   204600   
7748      7749  131400.0  131600.0  131800.0  ...   265700   267800   269200   
9171      9172  133100.0  133200.0  133300.0  ...   256200   257200   258800   

      2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
1      314100   315000   316600   318100   319600   321100   321800  
34     191400   193000   194300   195200   196700   199100   201100  
52     418000   421600   423000   422800   424000   427600   430200  
81     230600   233100   234900   235900   237700   240300   241800  
153    259600   260200   261100   262200   264100   266500   267800  
...       ...      ...      ...      ...      ...      ...      ...  
7008   233900   234000   231500   229800   231000   233600   234600  
7144   178800   182800   187600   191200   194000   196400   197800  
7426   206900   209200   210600   211000   211900   214500   216600  
7748   270800   270900   270400   270100   271600   274100   275500  
9171   261700   265400   269400   273200   276200   279800   283700  

[66 rows x 272 columns]

The following school districts all of which serve six zipcodes look like the best canidates to produce investment recommendations from using comparitive home value prices of neighboring zipcode: 
* Hurst-Euless-Bedford ISD 
* Lewisville ISD
* Plano ISD  

In [139]:
#create new data frames for each school district  

#hurst-euless-bedford isd
heb = dallas.loc[(dallas['RegionName'] == 76039) | (dallas['RegionName'] == 76021) | (dallas['RegionName'] == 76053)
                | (dallas['RegionName'] == 76040) | (dallas['RegionName'] == 76054) | (dallas['RegionName'] == 76022)] 


#lewisville isd 
lewis = dallas.loc[(dallas['RegionName'] == 75067) | (dallas['RegionName'] == 75056) | (dallas['RegionName'] == 75028)
                | (dallas['RegionName'] == 75077) | (dallas['RegionName'] == 75010) | (dallas['RegionName'] == 75022)
                  | (dallas['RegionName'] == 75057)]

#plano isd  
plano = dallas.loc[(dallas['RegionName'] == 75093) | (dallas['RegionName'] == 75025) | (dallas['RegionName'] == 75074)
                | (dallas['RegionName'] == 75023) | (dallas['RegionName'] == 75075) | (dallas['RegionName'] == 75252)
                  | (dallas['RegionName'] == 75082)]

## 4 - Reshape from Wide to Long Format

In [108]:
#pregiven helper function to melt dataframe
def melt_data(df):
    melted = pd.melt(df, id_vars=['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', 'SizeRank'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value':'mean'})

In [145]:
#melt inital 3 school grouped dataframes
long_heb = melt_data(heb) 
long_lewis = melt_data(lewis) 
long_plano = melt_data(plano)

## 5 - ARIMA Modeling

In [ ]:
def get_datetimes(df):
    return pd.to_datetime(df.columns.values[1:], format='%Y-%m')

## 6 - Conclusion & Recommendations 

In [ ]:
test.columns.to_list()

In [ ]:
dropped_melt

In [ ]:
dropped_melt.set_index('variable', inplace=True)

In [ ]:
dropped_melt.plot(figsize = (22,8))
plt.show()

# Step 6: Interpreting Results